WebApp/Utils/MSALPerUserMemoryTokenCache.cs

```diff
@@ -0,0 +1,198 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+***********************************************************************************************/
+
+using Microsoft.Identity.Client;
+using System;
+using System.Runtime.Caching;
+using System.Security.Claims;
+
+namespace WebApp.Utils
+{
+    /// <summary>
+    /// An implementation of token cache for both Confidential and Public clients backed by MemoryCache.
+    /// MemoryCache is useful in Api scenarios where there is no HttpContext to cache data.
+    /// </summary>
+    /// <seealso cref="https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/token-cache-serialization"/>
+    public class MSALPerUserMemoryTokenCache
+    {
+        /// <summary>
+        /// The backing MemoryCache instance
+        /// </summary>
+        internal readonly MemoryCache memoryCache = MemoryCache.Default;
+
+        /// <summary>
+        /// The duration till the tokens are kept in memory cache. In production, a higher value, upto 90 days is recommended.
+        /// </summary>
+        private readonly DateTimeOffset cacheDuration = DateTimeOffset.Now.AddHours(48);
+
+        /// <summary>
+        /// The internal handle to the client's instance of the Cache
+        /// </summary>
+        private ITokenCache UserTokenCache;
+
+        /// <summary>
+        /// Once the user signes in, this will not be null and can be obtained via a call to Thread.CurrentPrincipal
+        /// </summary>
+        internal ClaimsPrincipal SignedInUser;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MSALPerUserMemoryTokenCache"/> class.
+        /// </summary>
+        /// <param name="tokenCache">The client's instance of the token cache.</param>
+        public MSALPerUserMemoryTokenCache(ITokenCache tokenCache)
+        {
+            this.Initialize(tokenCache, ClaimsPrincipal.Current);
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="MSALPerUserMemoryTokenCache"/> class.
+        /// </summary>
+        /// <param name="tokenCache">The client's instance of the token cache.</param>
+        /// <param name="user">The signed-in user for whom the cache needs to be established.</param>
+        public MSALPerUserMemoryTokenCache(ITokenCache tokenCache, ClaimsPrincipal user)
+        {
+            this.Initialize(tokenCache, user);
+        }
+        
+        /// <summary>Initializes the cache instance</summary>
+        /// <param name="tokenCache">The ITokenCache passed through the constructor</param>
+        /// <param name="user">The signed-in user for whom the cache needs to be established..</param>
+        private void Initialize(ITokenCache tokenCache, ClaimsPrincipal user)
+        {
+            this.SignedInUser = user;
```

Comment:
>this. [](start = 12, length = 5)

nit: can this. be removed everywhere? just cleans things up a bit

Reply:
I belive that the coding style in the sample mandates using `this.` - we use a different style in MSAL.

Reply:
All `this.` were removed. They were there because on ADAL version, this class would extend TokenCache so the `this.` were necessary.